### ISSUE TITLE
Enh Req: for helidon-maven-plugin execution allow passing arbitrary jar for native-image

### ISSUE THREAD

**Issue by andrejusc** [oracle/helidon-build-tools#691]
Hi,

Within multi-module Maven project I have usage of helidon-maven-plugin defined within one of parent profiles. And within child module I have within build `finalName` usage like such:
```
  <build>
    <finalName>myservice</finalName>
    <plugins>
...
    </plugins>
  </build>
```
And in profile I have usage of helidon-maven-plugin like such to try quickly something:
```
        <plugins>
          <plugin>
            <groupId>io.helidon.build-tools</groupId>
            <artifactId>helidon-maven-plugin</artifactId>
            <executions>
              <execution>
                <goals>
                  <goal>native-image</goal>
                </goals>
                <configuration>
                  <finalName>aaa</finalName>
                </configuration>
              </execution>
            </executions>
          </plugin>

```
When invoking Maven with -X switch and with profile activated - in Debug output I could see such part:
```
[DEBUG] Configuring mojo 'io.helidon.build-tools:helidon-maven-plugin:2.2.1:native-image' with basic configurator -->
...
[DEBUG]   (f) execMode = jar
[DEBUG]   (f) finalName = aaa

```
But for native-image execution command I'm still having myservice.jar presence:
```
[DEBUG] Executing command: [/opt/graalvm-ce-java17-22.0.0.2/bin/native-image ... -jar, /dev/parent-module/child-module/target/myservice.jar
```
So question then - how for specific plugin execution I could override that jar name? I.e. I'd like it to take in my case `aaa.jar` as input and not `myservice.jar`

Looking into code:
https://github.com/oracle/helidon-build-tools/blob/936b0dcd84580b1a5bc6e114c8e6391ec81c3a29/maven-plugins/helidon-maven-plugin/src/main/java/io/helidon/build/maven/GraalNativeMojo.java#L110

I could see that `finalName` parameter for plugin is defined with such clause `readonly = true`

and that jar to be taken as input is calculated on plugin initialization potentially here:
https://github.com/oracle/helidon-build-tools/blob/936b0dcd84580b1a5bc6e114c8e6391ec81c3a29/maven-plugins/helidon-maven-plugin/src/main/java/io/helidon/build/maven/GraalNativeMojo.java#L578
and this part:
```
                    artifact = new File(buildDirectory,
                                        project.getBuild().getFinalName() + ".jar");
```
which is fine if some child module drives that name (originally above for my use case), but doesn't allow easy override within plugin execution without changing child POM, which I also need for more flexibility.

I've also tried to see if I could to achieve same by trying such:
```
                <configuration>
                  <additionalArgs>
                    <additionalArg>-jar</additionalArg>
                    <additionalArg>/dev/parent-module/child-module/target/aaa.jar</additionalArg>
                  </additionalArgs>
                </configuration>
```
but such then generates native-image execution command with 2 `-jar` options for 'aaa.jar' and 'myservice.jar', which I don't think then right way.

**Comment by andrejusc** [comment]
In my initial post I've noticed usage of outdated `v2.2.1` for `helidon-maven-plugin`. I've now retested with latest `v2.3.3` and so far same result.

**Comment by romain-grecourt** [comment]
Can you share the whole command line from the debug output ? The injected `finalName` field is used for `-H:Name=` ; is it different from the value used for `-jar` ?

I'm guessing that you the mojo execution fails right, can you confirm.

**Comment by andrejusc** [comment]
@romain-grecourt - so related parts of execution command for my use case are such:
```
[DEBUG] Executing command: [/opt/graalvm-ce-java17-22.0.0.2/bin/native-image ...,  -H:Name=aaa, ..., -jar, /dev/parent-module/child-module/target/myservice.jar]
```
From such I could say that one value is taken from within plugin execution configuration and another - still coming from my child module.

**Comment by romain-grecourt** [comment]
Thanks, this means the injected value for `finalName` should be use instead of `project.getBuild().getFinalName()`.

**Comment by romain-grecourt** [comment]
Actually, this is not a bug but like you've stated an enhancement.
The current version of the plugin always assume the main project artifact to be the jar file to use, it can be made configureable provide some flexibility.

Two changes:
- Make `finalName` non read-only to allow overriding the value from user configuration
- Expose a new configurable attribute `jarFile` that defaults to `project.getArtifact().getFile()`

**Comment by andrejusc** [comment]
@romain-grecourt - what is issue handling workflow here? I'm requestor here, but this is now closed. By what means should I validate and confirm that delivered change reflects the ask?

**Comment by andrejusc** [comment]
@romain-grecourt - may you have some interim snapshot build's jars somewhere available, so I could put those into my local Maven repo to test quickly instead of doing local build? Also - what is expected release target date for this issue to be included?

**Comment by romain-grecourt** [comment]
We do not deploy snapshots.
Please verify the changes, once you've confirmed everything works I can start working on the release.

**Comment by andrejusc** [comment]
I've built now helidon-maven-plugin-2.0.0-SNAPSHOT.jar from requested `helidon-2.x` branch. But if switching to it in my pom.xml - getting such problem:
```
ERROR] Failed to execute goal io.helidon.build-tools:helidon-maven-plugin:2.0.0-SNAPSHOT:native-image (default) on project openapi-repository: Image generation failed, exit code: 1 -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal io.helidon.build-tools:helidon-maven-plugin:2.0.0-SNAPSHOT:native-image (default) on project test-helidon: Image generation failed, exit code: 1
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:306)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:211)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:157)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:121)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:127)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:47)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:156)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:72)
Caused by: org.apache.maven.plugin.MojoFailureException: Image generation failed, exit code: 1
    at io.helidon.build.maven.graal.GraalNativeMojo.execute (GraalNativeMojo.java:265)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:301)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:211)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:157)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:121)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:127)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:47)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:156)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:72)
```
and such is actual issue:
```
[INFO] GraalVM Native Image: Generating ...
...
[WARNING] Fatal error: java.lang.NoClassDefFoundError: io/helidon/config/spi/ConfigSource
[WARNING] 	at java.base/java.lang.Class.getDeclaredMethods0(Native Method)
[WARNING] 	at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3402)
[WARNING] 	at java.base/java.lang.Class.getDeclaredMethod(Class.java:2673)
[WARNING] 	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:378)
[WARNING] 	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:587)
[WARNING] 	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:126)
[WARNING] 	at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:617)
[WARNING] Caused by: java.lang.ClassNotFoundException: io.helidon.config.spi.ConfigSource
[WARNING] 	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:445)
[WARNING] 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:587)
[WARNING] 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
[WARNING] 	... 7 more
[WARNING] Error: Image build request failed with exit status 1
```
So not yet exactly sure what is causing this and everything working fine if I switch back to 2.3.3 of helidon-mave-plugin.

**Comment by andrejusc** [comment]
@romain-grecourt - Ignore my above comment - I messed some folders for my test project.

I could confirm that having now such part inside pom.xml:
```
          <plugin>
            <groupId>io.helidon.build-tools</groupId>
            <artifactId>helidon-maven-plugin</artifactId>
            <version>2.0.0-SNAPSHOT</version>
            <configuration>
              <jarFile>${basedir}/target/mycustom.jar</jarFile>
              <finalName>myservice</finalName>
            </configuration>
          </plugin>
```
initially I could see such:
```
[DEBUG] Configuring mojo io.helidon.build-tools:helidon-maven-plugin:2.0.0-SNAPSHOT:native-image from plugin realm ClassRealm[plugin>io.helidon.build-tools:helidon-maven-plugin:2.0.0-SNAPSHOT, parent: java.net.URLClassLoader@23ab930d]
...
[DEBUG]   (f) execMode = jar
[DEBUG]   (f) finalName = myservice
[DEBUG]   (f) jarFile = /dev/target/mycustom.jar
...
```
and also such:
```
[DEBUG] Executing command: [/opt/graalvm-ce-java17-22.0.0.2/bin/native-image, ..., -H:Name=myservice, ..., -jar, /dev/target/mycustom.jar]
```

But now I could see such:
```
[INFO] GraalVM Native Image: Generating 'mycustom'...
...
[INFO] Finished generating 'mycustom' in 4m 57s.
```

So question then - why generated native image doesn't have `finalName` (from above - should be myservice), but has `jarFile` name without extension?

**Comment by andrejusc** [comment]
@romain-grecourt - I suggest we reopen this issue based on my last comment as overall behaviour is not yet right. But closer now to what I was asking about.

**Comment by romain-grecourt** [comment]
Can you figure out what you want by using `native-image` directly ?

**Comment by andrejusc** [comment]
I'd like to avoid using `native-image` directly for my use case, so Maven could be utilized as general approach for various executions.

**Comment by romain-grecourt** [comment]
Let me rephrase. Can you find out the set of arguments you need with native-image for it to satisfy your use case?
From what I can see the plugin is working as intended. I.e. the plugin configuration and the native-image command line match.
